torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2357,103 +2357,325 @@ Tensor slice_backward_wrapper(
+
+  // Optimisation for svdvals: gA = U @ diag(gS) @ Vh
+  if (!gU.defined() && !gVh.defined()) {
+    return m >= n ? at::matmul(U, gS.unsqueeze(-1) * Vh)
+                  : at::matmul(U * gS.unsqueeze(-2), Vh);
+  }
+  // At this point, at least one of gU, gVh is defined
+
+  const bool is_complex = U.is_complex();
+  const auto UhgU = gU.defined() ? at::matmul(U.mH(), gU) : Tensor{};
+  const auto VhgV = gVh.defined() ? at::matmul(Vh, gVh.mH()) : Tensor{};
+
+  // Check for the invariance of the loss function, i.e.
+  // Im(diag(U^H gU)) + Im(diag(V^H gV)) = 0
+  if (is_complex) {
+    const auto imdiag_UhgU = gU.defined() ? at::imag(UhgU.diagonal(0, -2, -1))
+                                          : at::zeros_like(S);
+    const auto imdiag_VhgV = gVh.defined() ? at::imag(VhgV.diagonal(0, -2, -1))
+                                           : at::zeros_like(S);
+    // Rather lax atol and rtol, as we don't want false positives
+    TORCH_CHECK(at::allclose(imdiag_UhgU, -imdiag_VhgV, /*rtol=*/1e-2, /*atol=*/1e-2),
+                "svd_backward: The singular vectors in the complex case are specified up to multiplication "
+                "by e^{i phi}. The specified loss function depends on this phase term, making "
+                "it ill-defined.");
+  }
+
+  // gA = (skew(U^H gU) / E) S +  S ((skew(V^H gV) / E) + I o (gS + i Im(diag(U^H gU)) / S)
+  Tensor gA = [&] {
+    // ret holds everything but the diagonal of gA
+    auto ret = [&] {
+      const auto skew = [](const Tensor& A) { return A - A.mH(); };
+
+      const auto E = [&S]{
+        const auto S2 = S * S;
+        auto ret = S2.unsqueeze(-2) - S2.unsqueeze(-1);
+        // Any number a != 0 would, as we are just going to use it to compute 0 / a later on
+        ret.diagonal(0, -2, -1).fill_(1);
+        return ret;
+      }();
+
+      if (gU.defined()) {
+        if (gVh.defined()) {
+          return (skew(UhgU) * S.unsqueeze(-2) + S.unsqueeze(-1) * skew(VhgV)) / E;
+        } else {
+          return (skew(UhgU) / E) * S.unsqueeze(-2);
+        }
+      } else { // gVh.defined();
+        return S.unsqueeze(-1) * (skew(VhgV) / E);
+      }
+    }();
+    // Fill the diagonal
+    const auto diag = ret.diagonal(0, -2, -1);
+    if (is_complex && gU.defined() && gVh.defined()) {
+      if (gS.defined()) {
+        at::real(diag).copy_(gS);
+      } else {
+        // Not strictly necessary, but we do it for good measure
+        at::real(diag).zero_();
+      }
+      at::imag(diag).copy_(at::imag(UhgU.diagonal(0, -2, -1)) / S);
+    } else {
+      if (gS.defined()) {
+        diag.copy_(gS);
+      } else {
+        // Not strictly necessary, but we do it for good measure
+        diag.zero_();
+      }
+    }
+    return ret;
+  }();
+
+  if (m > n && gU.defined()) {
+    // gA = [UgA + (I_m - UU^H)gU S^{-1}]V^H
+    gA  = at::matmul(U, gA);
+    const auto gUSinv = gU / S.unsqueeze(-2);
+    gA = gA + gUSinv - at::matmul(U, at::matmul(U.mH(), gUSinv));
```

Comment:
Just to avoid any confusion. Do I understand that correctly that rank of U and V is at most `min(m, n)` due to the narrowing?

Reply:
If by rank you mean the dimension of the image, then, since `U` and `V` have orthogonal columns they are always full rank.

Reply:
No, I mean the number of vectors.

Reply:
If `m > n`, `U` is of size `m x n` and `V` of `n x n`.
If `n > m`, `U` is of size `m x m`  and `V` is of size `n x m`.
I mean, the usual stuff in a compact SVD, one of the two matrices (if you consider V^H) is the size of A and the other one is square and small.

Reply:
So, exactly as I thought, at most `min(m, n)`. Because of `full_matrices=True` by default I thought that V and U are the bases and hence had this confusion with probably inefficient matrix product associations. I wonder why would we need `full_matrices=True` by default anyway?

Reply:
That default is absolute nonsense. I tried to fight to get it changed in https://github.com/data-apis/array-api/issues/214 but you can see how the discussion went...

Reply:
I believe having `full_matrices=True` is not useful, and, moreover, it inflicts an unnecessary memory penalty for super wide/tall matrices.

Reply:
I see, we have to carry all those bases for super tall/wide matrices and suffer the consequences of low-memory :)

Reply:
Note that these matrices are trimmed in `derivatives.yaml` before calling this funciton if `full_matrices=True`, so it's fine. The main problem is that, since it's the default, if a user is unwary, doing `u, s, vh = torch.linalg.svd(A)` will give them unnecessarily large matrices.

Reply:
But they are trimmed with a narrow, right? This one is a view, so I am not sure it saves us...

Reply:
We could use a `clone()`, but I think it's better just to optimise for the case where `full_matrices=False`, as it's the one most people will use.

Reply:
Nothing we can do, if it is computed by the forward, the memory is already allocated so we might just use a view in the backward :)